tls/s2n_record_read.c

```diff
@@ -34,15 +34,21 @@ int s2n_sslv2_record_header_parse(
         uint8_t *client_protocol_version,
         uint16_t *fragment_length)
 {
-    struct s2n_stuffer *in = &conn->header_in;
+    struct s2n_stuffer *header_in = &conn->header_in;
 
-    S2N_ERROR_IF(s2n_stuffer_data_available(in) < S2N_TLS_RECORD_HEADER_LENGTH, S2N_ERR_BAD_MESSAGE);
+    POSIX_ENSURE(s2n_stuffer_data_available(header_in) >= S2N_TLS_RECORD_HEADER_LENGTH,
+            S2N_ERR_BAD_MESSAGE);
 
-    POSIX_GUARD(s2n_stuffer_read_uint16(in, fragment_length));
+    POSIX_GUARD(s2n_stuffer_read_uint16(header_in, fragment_length));
+    /* The first bit of the length must be set to indicate SSLv2,
+     * so the raw fragment length must be adjusted.
+     */
+    *fragment_length ^= (S2N_TLS_SSLV2_HEADER_FLAG << 8);
```

Comment:
I wrote a bunch of unit tests for this in s2n_record_read_test, but still please sanity check my math :)

Reply:
The shift math is correct 😄. I'm just a bit concerned about the XOR: https://github.com/aws/s2n-tls/pull/4425#discussion_r1494869352